### PR TITLE
Add patch disc loss

### PR DESCRIPTION
**Pull request by gabrieltseng**
Adds a patch disc loss with tests

### CODE REVIEW

File: helios/train/loss.py [hunk 1]
```diff
@@ -77,10 +30,33 @@ def compute(self, predictions: Any, targets: Any, **kwargs: Any) -> float:
...
+            return rearrange(x, "b c h w d -> b (h w c) d")
+        elif x.dim() == 4:
+            # (B, C_G, T, D)
+            return rearrange(x, "b c t d -> b (t c) d")
+        elif x.dim() == 3:
+            # (B, C_G, D)
+            return x
+
+    @staticmethod
+    def _expand_and_reciprocate(t):
```
Comment (Hgherzog) [inline helios/train/loss.py:69]: Suggestion: add a doc string showing an example of this manipulation (this is from an AI so verify it)
```
# Example transformation:
# If t = tensor([2, 3])
# reciprocals = tensor([0.5000, 0.3333])  # 1/2, 1/3
# final output = tensor([0.5000, 0.5000, 0.3333, 0.3333, 0.3333])
#               repeats each value t[i] times: 0.5 repeated 2x, 0.3333 repeated 3x
```

File: helios/train/loss.py [hunk 2]
```diff
@@ -94,11 +70,59 @@ def compute(
...
+                self._flatten(getattr(predictions, f"{d}_mask").unsqueeze(dim=-1))
+                for d in predictions.data_fields
+            ],
+            dim=1,
+        )[:, :, 0]
+        all_targets = torch.cat(
+            [self._flatten(getattr(targets, d)) for d in predictions.data_fields], dim=1
+        )
+
+        pred = all_preds[all_masks == 2].unsqueeze(dim=0)
```
Comment (Hgherzog) [inline helios/train/loss.py]: We should have an Enum for the different mask values and there meanings for readability and consistency

File: helios/train/loss.py [hunk 3]
```diff
@@ -77,10 +30,33 @@ def compute(self, predictions: Any, targets: Any, **kwargs: Any) -> float:
 class PatchDiscriminationLoss(Loss):
     """Loss function for patch discrimination task."""
 
-    def __init__(self, tau: float = 0.07, pred2unit: bool = True):
+    def __init__(
+        self, tau: float = 0.07, pred2unit: bool = True, mask_other_samples: bool = True
+    ):
         """Initialize patch discrimination loss."""
```
Comment (Hgherzog) [inline helios/train/loss.py]: add args to doc string

File: helios/train/loss.py [hunk 4]
```diff
@@ -94,11 +70,59 @@ def compute(
...
+            pred = (pred - pred_mu) / (pred_std + 1e-4)
+
+        pred = F.normalize(pred, p=2, dim=-1)
+        target = F.normalize(target, p=2, dim=-1)
+
+        scores = torch.einsum("npd,nqd->npq", pred, target) / self.tau
+        count = (all_masks == 2).sum(dim=-1)
+
+        if self.mask_other_samples:
+            logit_mask = torch.full_like(scores, -torch.finfo(scores.dtype).max)
```
Comment (Hgherzog) [inline helios/train/loss.py:121]: Suggestion: add a helper function for the mask_other_samples

File: helios/train/loss.py [hunk 5]
```diff
@@ -77,10 +30,33 @@ def compute(self, predictions: Any, targets: Any, **kwargs: Any) -> float:
 class PatchDiscriminationLoss(Loss):
```
Comment (Hgherzog) [inline helios/train/loss.py:31]: What we are calling patch discrimination loss here seems to be the AllDisc loss from the Galileo paper where we also average across the batch dimension? Maybe add a reference in the function doc string explaining this?

File: tests/unit/train/test_loss.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        latlon_mask=torch.ones((b, 1)) * 2,
+    )
+    targets = TokensAndMasks(
+        s2=torch.ones((b, t, t_h, t_w, d)),
+        s2_mask=torch.zeros((b, t, t_h, t_w)),
+        latlon=torch.ones((b, 1, d)),
+        latlon_mask=torch.zeros((b, 1)),
+    )
+    loss = PatchDiscriminationLoss()
+    _ = loss.compute(preds, targets)
```
Comment (Hgherzog) [inline tests/unit/train/test_loss.py]: Suggestion: add an assert for the loss value, at least that it is approximately what we expect (checking full precision will probably create more trouble than it is worth

File: helios/train/loss.py (hunk 2 above)
Comment (gabrieltseng) [reply helios/train/loss.py]: Added.

File: helios/train/loss.py (hunk 1 above)
Comment (gabrieltseng) [reply helios/train/loss.py:69]: Added.

File: helios/train/loss.py (hunk 3 above)
Comment (gabrieltseng) [reply helios/train/loss.py]: Added

File: helios/train/loss.py (hunk 5 above)
Comment (gabrieltseng) [reply helios/train/loss.py:31]: Added

File: tests/unit/train/test_loss.py (hunk 6 above)
Comment (gabrieltseng) [reply tests/unit/train/test_loss.py]: Added

File: helios/train/loss.py (hunk 4 above)
Comment (gabrieltseng) [reply helios/train/loss.py:121]: I am reluctant to excessively nest this function, so would prefer to keep it as is. If you feel strongly, I'll change it
